### PR TITLE
feat: allow user deletion without deleting task references

### LINKED ISSUE #1661: Backend API to delete user account
**Is your feature request related to a problem? Please describe.**

- This is essential for GDPR. 

**Describe the solution you'd like**

- Deletion of a user db entry, in addition to keeping references to the user for task history in place.
- How to keep the references in place? Save the username and profile picture in the task history?

### PR DESCRIPTION
**Pull request by Anuj-Gupta4**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature

## Related Issue

Fixes #1661 

## Describe this PR

On user deletion, task history will retain username from user
and rest of the user data will be deleted.

## Checklist before requesting a review

- 📖 Read the FMTM Contributing Guide: <https://github.com/hotosm/fmtm/blob/main/CONTRIBUTING.md>
- 📖 Read the HOT Code of Conduct: <https://docs.hotosm.org/code-of-conduct>
- 👷‍♀️ Create small PRs. In most cases, this will be possible.
- ✅ Provide tests for your changes.
- 📝 Use descriptive commit messages.
- 📗 Update any related documentation and include any relevant screenshots.

### CODE REVIEW

File: src/backend/app/tasks/tasks_crud.py
Comment (Sujanadh) [inline src/backend/app/tasks/tasks_crud.py:178]: provide `Session` type to db also.

File: src/backend/app/tasks/tasks_crud.py
Comment (Sujanadh) [inline src/backend/app/tasks/tasks_crud.py:213]: instead of `fetchone() `simply use` first()`, `first()` will call only first row all the time, while `fetchone()` returns all the rows but only one at a time.

Comment (spwoodcock) [reply src/backend/app/tasks/tasks_crud.py:213]: This is true for SQLAlchemy, but not for psycopg actually!

Psycopg only has fetchone fetchmany and fetchall 😄

Apologies for not mentioning that!

**Comment by spwoodcock** [comment]
So sorry @Anuj-Gupta4 , but this PR is based on staging, which is using SQLAlchemy 😅

We have since done a huge refactor on development to remove this and use psycopg, so this PR might need to be started from scratch with development as the base 😥

File: src/backend/app/tasks/tasks_crud.py
Comment (spwoodcock) [reply src/backend/app/tasks/tasks_crud.py:213]: Actually you are technically correct!

I noticed this PR is using SQLAlchemy! Commented in the main thread

**Comment by spwoodcock** [comment]
There is some really nice code here 😄

Could you possibly rebase with development (might be a lot of hassle considering the amount of change), or stash your changes somewhere, then reset the branch and reapply?

Also feel free to make a new PR if it's easier, then I'll take a look!

**Comment by Anuj-Gupta4** [comment]
@spwoodcock I have reset my local branch and am working over development code. I will be overwriting the git history once its completed.

File: src/backend/app/db/models.py
Comment (spwoodcock) [inline src/backend/app/db/models.py:98]: Is there a use case for a UserRole entry with no user_id set?

As we won't know the user for which the role relates to

File: src/backend/app/db/models.py
Comment (spwoodcock) [inline src/backend/app/db/models.py:258]: Perfect 👍

File: src/backend/app/db/models.py
Comment (spwoodcock) [inline src/backend/app/db/models.py:558]: Same for this - I don't know when user_id would be empty for a record

File: src/backend/migrations/010-delete-user.sql
Comment (spwoodcock) [inline src/backend/migrations/010-delete-user.sql:16]: No need to specify max 255 chars - it has no efficiency gain in Postgres. Just VARCHAR is good 👍

**Review by spwoodcock — CHANGES_REQUESTED** [review]
Very nice work! 🙌

File: src/backend/app/db/models.py
Comment (Sujanadh) [inline src/backend/app/db/models.py:227]: Can we change this to a single SQL and await the execution at once? What do you think?

**Comment by Anuj-Gupta4** [comment]
@spwoodcock The tests are passing locally. I am not sure why it is failing here.

File: src/backend/app/users/user_routes.py
Comment (Sujanadh) [inline src/backend/app/users/user_routes.py:85]: I think it's best to return a response with a similar message as in the log instead of `NO_CONTENT`.

File: src/backend/app/db/models.py
Comment (spwoodcock) [reply src/backend/app/db/models.py:227]: Unfortunately that's not possible with psycopg! It's only one statement per cursor 😅
(SQLAlchemy hides this fact!).

It looks inefficient, but it still executes them all in the same session, so it's not bad

Comment (Sujanadh) [reply src/backend/app/db/models.py:227]: I see you have done a similar delete api for the organisation with a single sql.

Comment (spwoodcock) [reply src/backend/app/db/models.py:227]: Oh right! Does it actually work? I would think it might error, but could be wrong. 

I think the project deletion has it separated for this reason

Comment (spwoodcock) [reply src/backend/app/db/models.py:227]: Ah ignore me, I see what you mean now!

The org deletion is a single prepared statement, so that works fine

Comment (Sujanadh) [reply src/backend/app/db/models.py:227]: I am not sure if it works, I haven't tested it yet. :thinking:

Comment (spwoodcock) [reply src/backend/app/db/models.py:227]: > I am not sure if it works, I haven't tested it yet. :thinking: 

I think it does!

**Comment by spwoodcock** [comment]
Oops forgot to wait for the other PR comments to be addressed before merge. 

It would be nice to possibly address them in another PR if time @Anuj-Gupta4 🙏

**Pull request merged**
